### PR TITLE
estimate-area command 

### PR DESCRIPTION
**Pull request by allisonzheng**
The estimate-area command inputs features and a precision and outputs the total area in square kilometers.

### CODE REVIEW

**Comment by allisonzheng** [comment]
The build failures for some of the commits were before a merge with master with upgraded the Click version and fixed AssertionError quote issues.

File: mapbox_tilesets/scripts/cli.py [hunk 1]
```diff
@@ -703,3 +705,70 @@ def list_sources(username, token=None):
...
+    type=click.Choice(["10m", "1m", "30cm", "1cm"]),
+    help="Precision level",
+)
+@click.option(
+    "--no-validation",
+    required=False,
+    is_flag=True,
+    help="Bypass source file validation",
+)
+@click.option("--force-1cm", required=False, is_flag=True, help="Enables 1cm precision")
```
Comment (ericcarlschwartz) [inline mapbox_tilesets/scripts/cli.py:725]: the formatting on this option doesn't match the two above. 

could we update it to match OR do we have a linter that we use on this repo?

File: mapbox_tilesets/scripts/cli.py [hunk 2]
```diff
@@ -703,3 +705,70 @@ def list_sources(username, token=None):
...
+    try:
+        features = builtins.list(filter_features(features))
+    except (ValueError, json.decoder.JSONDecodeError):
+        raise errors.TilesetsError(
+            "Error with feature parsing. Ensure that feature inputs are valid and formatted correctly. Try 'tilesets estimate-area --help' for help."
+        )
+    except Exception:
+        raise errors.TilesetsError("Error with feature filtering.")
+
+    if not no_validation:
```
Comment (ericcarlschwartz) [inline mapbox_tilesets/scripts/cli.py:753]: in what circumstances would we expect/want users to bypass source validation?

File: mapbox_tilesets/scripts/cli.py [hunk 3]
```diff
@@ -703,3 +705,70 @@ def list_sources(username, token=None):
...
+        for feature in features:
+            utils.validate_geojson(feature)
+
+    try:
+        area = utils.calculate_tiles_area(features, precision)
+    except Exception:
+        raise errors.TilesetsError(
+            "Error with are calculation. Please check features and try again."
```
Comment (ericcarlschwartz) [inline mapbox_tilesets/scripts/cli.py:761]:
```suggestion
            "Error with area calculation. Please check features and try again."
```

**Comment by maialena** [comment]
Hey Allison! I haven't had a chance to do a full review (have bookmarked this PR for later this afternoon) but do you mind linking the original ticket for this work so I can learn more about the context?

File: mapbox_tilesets/utils.py [hunk 4]
```diff
@@ -103,3 +108,59 @@ def validate_geojson(feature):
...
+    return ((tile_x / 2 ** zoom) * 360.0) - 180.0
+
+
+def _tile2lat(tile_y, zoom):
+    """Returns tile latitude"""
+    n = np.pi - 2 * np.pi * tile_y / 2 ** zoom
+    return (180.0 / np.pi) * np.arctan(0.5 * (np.exp(n) - np.exp(-n)))
+
+
+def _calculate_tiles_area(tiles):
```
Comment (ericcarlschwartz) [inline mapbox_tilesets/utils.py:136]: do we want to rename this to make clear it only returns the area for a single tile?

File: tests/test_cli_estimate_area.py [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+
+def test_cli_estimate_area_with_no_validation_flag_and_invalid_feature():
+    runner = CliRunner()
+    # the input is missing "properties"
+    validated_result = runner.invoke(
+        estimate_area,
+        ["--precision", "10m", "--no-validation"],
+        input='{"type": "Feature", "geometry": { "type": "Point","coordinates": [125.6, 10.1]}}',
+    )
+    assert validated_result.exit_code == 0
```
Comment (ericcarlschwartz) [inline tests/test_cli_estimate_area.py:153]: why is there no assert on the output here? trying to understand what we return to users when they skip validation and have invalid geojson

File: tests/test_cli_estimate_area.py [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+    message = "Error with feature parsing. Ensure that feature inputs are valid and formatted correctly. Try 'tilesets estimate-area --help' for help."
+    runner = CliRunner()
+    invalidated_result = runner.invoke(
+        estimate_area, ["--precision", "10m"], input="invalidGeoJson input"
+    )
+    assert invalidated_result.exit_code == 1
+    assert str(invalidated_result.exception) == message
+
+
+def test_cli_estimate_area_features_from_invalid_geojson_content():
```
Comment (ericcarlschwartz) [inline tests/test_cli_estimate_area.py:17]: does this particular error case not return a message?

**Review by ericcarlschwartz — COMMENTED** [review]
hey @allisonzheng, looks great! just had a couple questions to help myself understand what's going on here and a couple small suggestions

**Comment by allisonzheng** [comment]
Hi Maia,

Yeah! This is the ticket
<https://github.com/mapbox/tilesets-api-team/issues/419>. Thanks for
reviewing it! :D

File: tests/test_cli_estimate_area.py (hunk 5 above)
Comment (allisonzheng) [reply tests/test_cli_estimate_area.py:153]: The validation that's being skipped is that a user's features follows the GeoJSON schema correctly (has type, geometry and properties) but it would throw earlier (line 751) if it was missing features since we extract the features first. But it's missing an output assertion I'll add that in!

File: mapbox_tilesets/scripts/cli.py (hunk 2 above)
Comment (allisonzheng) [reply mapbox_tilesets/scripts/cli.py:753]: Added a comment in the code but when users rerun the command and know that their features already passed validation previously.

File: mapbox_tilesets/utils.py [hunk 7]
```diff
@@ -103,3 +108,59 @@ def validate_geojson(feature):
...
+    ----------
+    features: list
+        features from GeoJSON source and coordinates
+
+    Returns
+    -------
+        total area of all tiles in square kilometers
+    """
+    zoom = _convert_precision_to_zoom(precision)
+    tiles = burn(features, min(zoom, MAXZOOM_TILE_COVER))
```
Comment (dnomadb) [inline mapbox_tilesets/utils.py:165]: See above note about `MAXZOOM_TILE_COVER
```suggestion
    tiles = burn(features, zoom)
```

File: mapbox_tilesets/scripts/cli.py [hunk 8]
```diff
@@ -703,3 +705,76 @@ def list_sources(username, token=None):
...
+
+    # expect users to bypass source validation when users rerun command and their features passed validation previously
+    if not no_validation:
+        for feature in features:
+            utils.validate_geojson(feature)
+
+    try:
+        area = utils.calculate_tiles_area(features, precision)
+    except Exception:
+        raise errors.TilesetsError(
```
Comment (dnomadb) [inline mapbox_tilesets/scripts/cli.py:766]: I think that we should remove the try / except unless there are specific errors we want to catch / message in a specific way. Allowing the original error to raise should allow better communication with users for debugging.

File: tests/test_utils.py [hunk 9]
```diff
@@ -65,3 +66,23 @@ def test_validate_tileset_id_toolong():
...
+
+
+def test_convert_precision_to_zoom_30cm():
+    precision = "30cm"
+    assert _convert_precision_to_zoom(precision) == 14
+
+
+def test_convert_precision_to_zoom_1cm():
+    precision = "1cm"
+    return _convert_precision_to_zoom(precision) == 17
```
Comment (dnomadb) [inline tests/test_utils.py:88]: Since you've modularized the scripts in `mapbox_tilesets/utils.py`, I think we could have many of the CLI cases as unit tests; will be easier to assert on as opposed to CLI tests.

**Review by dnomadb — CHANGES_REQUESTED** [review]
@allisonzheng looks good! I added a few suggestions for changes. My biggest comment is that we should focus on unit testing the methods you've written instead of CLI tests -- they are harder to write and more prone to weirdness with things like string escaping. Some CLI tests should exist for usage, but the coverage of logical edge cases should be in module tests.

**Review by ericcarlschwartz — APPROVED** [review]
Thanks for addressing my comments/questions!

File: mapbox_tilesets/scripts/cli.py [hunk 10]
```diff
@@ -703,3 +705,70 @@ def list_sources(username, token=None):
...
+    """Estimate area of features with a precision level.
+
+    tilesets estimate-area <features> <precision>
+
+    features must be a list of paths to local files containing GeoJSON feature collections or feature sequences from argument or stdin, or a list of string-encoded coordinate pairs of the form "[lng, lat]", or "lng, lat", or "lng lat".
+    """
+    area = 0
+    if precision == "1cm" and not force_1cm:
+        raise errors.TilesetsError(
+            "The --force-1cm flag must be present to enable 1cm precision area calculation and may take longer for large feature inputs. 1cm precision for tileset processing is only available upon request after contacting Mapbox support."
```
Comment (dnomadb) [inline mapbox_tilesets/scripts/cli.py:741]:
```suggestion
            "The --force-1cm flag must be present to enable 1cm precision area calculation and may take longer for large feature inputs or data with global extents. 1cm precision for tileset processing is only available upon request after contacting Mapbox support."
```

**Review by dnomadb — APPROVED** [review]
Looks good! Just left one small documentation suggestion.

🙇  @allisonzheng !

**Comment by allisonzheng** [comment]
**Additional Testing** 
Using Mode the 99th percentile for upload sizes is ~243MB in the last year 
Screen Shot 2020-10-15 at 2 46 20 PM

I uploaded a ~440MB geonjson file converted from US census data and verified that the CLI processed it and calculated the area correctly 

**Mapbox Studio**
Screen Shot 2020-10-15 at 2 59 34 PM
**Tilesets CLI**
Screen Shot 2020-10-15 at 2 59 55 PM
